### PR TITLE
[Docs] Add migration notes for the high-level rest client

### PR DESCRIPTION
**Pull request by tlrx**
This PR adds some notes in the Java API and Reference documentation that mention the existence of the Java High Level Rest Client.

It also adds a migration page in the high level client documentation. It is definitively perfectible but that's a start.

### CODE REVIEW

File: docs/java-rest/high-level/apis/main.asciidoc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+=== Info API
+
+[[java-rest-high-main-request]]
+==== Execution
+
+Cluster information can be retrieved using the `info()` method:
+
+["source","java",subs="attributes,callouts,macros"]
+--------------------------------------------------
+include-tagged::{doc-tests}/CRUDDocumentationIT.java[main-execute]
```
Comment (nik9000) [inline docs/java-rest/high-level/apis/main.asciidoc]: It isn't really a Create, Read, Update, or Delete operation so maybe it should be in its own test just for naming's sake.

File: docs/java-rest/high-level/migration.asciidoc [hunk 2]
```diff
@@ -0,0 +1,137 @@
+[[java-rest-high-level-migration]]
+== Migration Guide
+
+This section describes how to migrate existing code from the `TransportClient`
+to the new Java High Level REST Client released with the version 5.6.0
+of Elasticsearch.
+
+=== Motivations around a new Java client
+
+The existing `TransportClient` has always been part of Elasticsearch since https://github
```
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: I'd s/always//.

File: docs/java-rest/high-level/migration.asciidoc [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+.com/elastic/elasticsearch/blob/b3337c312765e51cec7bde5883bbc0a08f56fb65/modules/elasticsearch/src/main/java/org/elasticsearch/client
+/transport/TransportClient.java[its very first commit]. Although it covers the whole set of Elasticsearch APIs,
+it always has had various flaws:
+
+- it is tightly coupled with Elasticsearch and uses Elasticsearch's internal implementations
+- it uses Java serialization to transport objects over the wire using their binary representations
```
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: It doesn't use *Java* serialization, right? I mean, it uses a our home grown binary serialization.

File: docs/java-rest/high-level/migration.asciidoc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+These two aspects forces any Java application that communicates with an Elasticsearch cluster to use
+a `TransportClient` in a version compatible with the remote cluster's version. That's perfect most of
```
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: I wouldn't say it is perfect - it is acceptable, but it ties the user application quite tightly to Elasticsearch.

File: docs/java-rest/high-level/migration.asciidoc [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+the time but Elasticsearch moves fast and new versions are released every month, forcing Java applications
+to use a newer version of the `TransportClient` just because the remote cluster has been upgraded.
+
+We decided to tackle these issues and we released a low-level REST client in 2016. This low-level REST client
+is based on the well-known Apache HTTP Client and it allows to communicate with an Elasticsearch cluster in any
+version using HTTP. On top of it, we released the high-level REST client. It is based on the low-level client
+and it provides a <<java-rest-high-supported-apis,subset of Elasticsearch's APIs>> that uses Elasticsearch core
+Java classes, like the `TransportClient`does, in order to perform requests and to parse responses.
```
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: missing space.

File: docs/java-rest/high-level/migration.asciidoc [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    <groupId>org.elasticsearch.client</groupId>
+    <artifactId>transport</artifactId>
+    <version>...</version>
+</dependency>
+--------------------------------------------------
+
+This dependency must be removed and replaced by a new dependency on the
+high-level client. The <<java-rest-high-usage,Getting Started>> page shows
+ typical configurations for Maven and Gradle and presents the
+ <<java-rest-high-usage-dependencies,two transitive dependencies>> brought
```
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: It has more than two transitive dependencies because it has Elasticsearch's dependencies too.

File: docs/java-rest/high-level/migration.asciidoc (hunk 4 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: I agree - I changed _perfect_ for _acceptable_. Let me know if you think the whole sentence should be reworded.

File: docs/java-rest/high-level/migration.asciidoc (hunk 6 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Oh right - I replaced _two transitive dependencies_ by _dependencies_ as I find it simpler to understand.

File: docs/java-rest/high-level/migration.asciidoc (hunk 3 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Absolutely... I had trouble to find the right phrase and in the end it ended with something wrong :/

**Comment by tlrx** [comment]
@nik9000 Thanks a lot for your feedback. I applied some changes, but I'd love to hear if you find this migration page useful or not or if you think something is missing here. I also think that @javanna has some ideas on it.

File: docs/java-rest/high-level/migration.asciidoc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+=== Prerequisite
+
+The Java High Level Rest Client requires Java `1.8`.
+
+=== Compatibility
+
+==== Java API
+The Java High Level Rest Client has been released with the version 5.6.0 of Elasticsearch and is
+therefore compatible with the Elasticsearch's Java API in this version.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: I wonder what you mean by compatible here. There are breaking changes, and that's why we are working on this page?

File: docs/java-rest/high-level/migration.asciidoc [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+==== Rest API
+A special effort has been made to make it compatible with the Elasticsearch's REST API in versions
+5.x and 6.x.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: what do you mean here?

File: docs/java-rest/high-level/migration.asciidoc [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+Changing an existing code to use the `RestHighLevelClient` instead of the `TransportClient`
+requires the following steps:
+
+- Update dependencies
+- Update client initialization
+- Update application code
+
+==== Updating the dependencies
+
+Java application that uses the `TransportClient` declare the following dependency:
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: declare**s**

File: docs/java-rest/high-level/migration.asciidoc [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+--------------------------------------------------
+
+Both `RestClient` and `RestHighLevelClient` are thread safe. They are
+ typically instantiated by the application at startup time or when the
+ first request is executed, and must both be closed when they are not
+ needed anymore or when the application is stopped.
+
+The `RestClient` provides various <<java-rest-low-usage-initialization,
+ initialization options>> that can be used to tune the way it executes
+ the requests.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: I think we can drop this sentence unless we want to add examples

File: docs/java-rest/high-level/migration.asciidoc [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+==== Changing the application's code
+
+The `RestHighLevelClient` does not provide the same methods than
+the `TransportClient` to execute the requests.
+
+More important, the high-level client:
+
+- does not provide indices or cluster management APIs. Management
+operations must be executed by external scripts or using the low-level
+client.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: does not provide (yet)  ? we may provide them at some point... would be nice to add an example of using the low-level client for some simple admin operation.

File: docs/java-rest/high-level/migration.asciidoc [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+- does not provide request builders like the `TransportClient`. The
+ legacy methods like `client.prepareIndex()` must be changed to use
+ request constructors like `new IndexRequest()` to create requests
+ instances. The requests are then executed using synchronous or
+ asynchronous dedicated methods like `client.index()` or `client.indexAsync()`.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: This is the most important point and I would expand on it, with examples of existing code and how it needs to become.

I would add one first example of java api using requests. sync and async, and their replacement.

Then I would expand on request builders, with some example on how to go from request builders to plain requests. Do we want to give people the hint to keep the transport dependency, create a dummy client, and do client.prepareIndex(index, type, id).request() to retrieve the request, which can be reused with the hl client? I am not sure as it is a temporary solution but it can save time in the beginning.

File: docs/java-rest/high-level/migration.asciidoc [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+the `TransportClient` to execute the requests.
+
+More important, the high-level client:
+
+- does not provide indices or cluster management APIs. Management
+operations must be executed by external scripts or using the low-level
+client.
+- does not provide request builders like the `TransportClient`. The
+ legacy methods like `client.prepareIndex()` must be changed to use
+ request constructors like `new IndexRequest()` to create requests
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: request objects? Also `Requests.indexRequest` is still valid. But maybe we don't want to promote it as it doesn't buy users anything.

File: docs/java-rest/high-level/migration.asciidoc [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+The `RestHighLevelClient` does not provide the same methods than
+the `TransportClient` to execute the requests.
+
+More important, the high-level client:
+
+- does not provide indices or cluster management APIs. Management
+operations must be executed by external scripts or using the low-level
+client.
+- does not provide request builders like the `TransportClient`. The
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: does not support. I would take out the "like the TransportClient" part.

File: docs/java-rest/high-level/migration.asciidoc [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+==== Changing the application's code
+
+The `RestHighLevelClient` does not provide the same methods than
+the `TransportClient` to execute the requests.
+
+More important, the high-level client:
+
+- does not provide indices or cluster management APIs. Management
+operations must be executed by external scripts or using the low-level
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: s/must/can ?

File: docs/java-rest/high-level/migration.asciidoc [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+The `RestClient` provides various <<java-rest-low-usage-initialization,
+ initialization options>> that can be used to tune the way it executes
+ the requests.
+
+Once the `RestHighLevelClient` is initialized, it can then be used to
+execute any of the <<java-rest-high-supported-apis,supported APIs>>.
+
+WARNING: The high-level client does not verify the name of the cluster
+it connects to. This verification can be made after the initialization
+of the client by checking the <<java-rest-high-main, cluster's information>>
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: I wonder if this warning is needed. Isn't this kind of implicit given that you never specify the cluster name when specifying the hosts?

File: docs/java-rest/high-level/migration.asciidoc [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+["source","xml",subs="attributes"]
+--------------------------------------------------
+<dependency>
+    <groupId>org.elasticsearch.client</groupId>
+    <artifactId>transport</artifactId>
+    <version>...</version>
+</dependency>
+--------------------------------------------------
+
+This dependency must be removed and replaced by a new dependency on the
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: must be replaced is enough?

File: docs/java-rest/high-level/migration.asciidoc [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+The Java High Level Rest Client has been released with the version 5.6.0 of Elasticsearch and is
+therefore compatible with the Elasticsearch's Java API in this version.
+
+==== Rest API
+A special effort has been made to make it compatible with the Elasticsearch's REST API in versions
+5.x and 6.x.
+
+=== How to migrate
+
+Changing an existing code to use the `RestHighLevelClient` instead of the `TransportClient`
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: Adapting existing code  ?

File: docs/java-rest/high-level/migration.asciidoc [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+These two aspects forces any Java application that communicates with an Elasticsearch cluster to use
+a `TransportClient` in a version compatible with the remote cluster's version. That's acceptable most of
+the time but Elasticsearch moves fast and new versions are released every month, forcing Java applications
+to use a newer version of the `TransportClient` just because the remote cluster has been upgraded.
+
+We decided to tackle these issues and we released a low-level REST client in 2016. This low-level REST client
+is based on the well-known Apache HTTP Client and it allows to communicate with an Elasticsearch cluster in any
+version using HTTP. On top of it, we released the high-level REST client. It is based on the low-level client
+and it provides a <<java-rest-high-supported-apis,subset of Elasticsearch's APIs>> that uses Elasticsearch core
+Java classes, like the `TransportClient` does, in order to perform requests and to parse responses.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: gut feeling: these motivations are more for a blogpost that for our docs maybe? I would rather focus on how to migrate here.

**Review by javanna — CHANGES_REQUESTED** [review]
thanks for working on this, I left some comments

File: docs/java-rest/high-level/migration.asciidoc (hunk 19 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: I don't know - I added this to add some context around why it's good to migrate, since it's easy to miss a blog post. But I can remove this whole section if you really feel it should not be here.

File: docs/java-rest/high-level/migration.asciidoc (hunk 16 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: By writing this, I wanted to "promote" what I think would be a good practice... on the other side, nothing prevent the user to configure multiple hosts from different clusters :( Let's remove this.

File: docs/java-rest/high-level/migration.asciidoc (hunk 13 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Yeah, I don't think `Requests` is very helpful so I didn't mention it here.

File: docs/java-rest/high-level/migration.asciidoc (hunk 19 above)
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: I like it. Maybe it should be in a blog post as well, but I like explaining it here as too.

File: docs/java-rest/high-level/migration.asciidoc (hunk 12 above)
Comment (nik9000) [inline docs/java-rest/high-level/migration.asciidoc]: > Do we want to give people the hint to keep the transport dependency, create a dummy client, and do client.prepareIndex(index, type, id).request() to retrieve the request, which can be reused with the hl client? I am not sure as it is a temporary solution but it can save time in the beginning.

I think it makes sense as an automated migration mechanism but I don't particularly like the idea of telling people to do it. Maybe something like:

While you are incrementally migrating your application and you have both the transport client and the high level client available you can always get the Request object from the Builder object by calling `Builder.request()`. We do not advise continuing to depend on the builders in the long run but it should be possible to use them during the transition from the transport client to the high level rest client.


I dunno. Now that I write that it just sounds like a bunch of words.

File: docs/java-rest/high-level/migration.asciidoc (hunk 7 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Yeah, I agree it's not clear. I rephrased that, I just wanted to say that you can use the client to connect to a cluster in version 5.6.0+.

File: docs/java-rest/high-level/migration.asciidoc (hunk 8 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: That you can send requests to a cluster in version 5.6.0+ too and that we relaxed the parsing logic to make it kind of forward compatible with possible changes in the REST API. But now I think it's better to just say that one can use it with any cluster in version 5.6.0+.

File: docs/java-rest/high-level/migration.asciidoc (hunk 19 above)
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: ok let's keep it

File: docs/java-rest/high-level/migration.asciidoc (hunk 12 above)
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: agreed, if we ought to mention this, we should do it the way you suggest (saying it's not advisable on the long run etc)

Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: I think what Nik suggested is ok and it can help people to migrate their applications.

**Comment by tlrx** [comment]
Thanks @javanna for your review, it is very helpful. I tried to complete the documentation following your indications. I'm pretty sure it still need some work but it gets closer, I hope.

File: docs/java-api/index.asciidoc [hunk 20]
```diff
@@ -17,6 +17,10 @@ Additionally, operations on a client may be accumulated and executed in
 Note, all the APIs are exposed through the
 Java API (actually, the Java API is used internally to execute them).
 
+WARNING: Starting from version 5.6.0, a new Java client has been
+released: the {java-rest}/java-rest-high.html[Java High Level REST Client].
+This new client is designed to replace the usage of the `TransportClient`
```
Comment (javanna) [inline docs/java-api/index.asciidoc]: to replace the TransportClient (remove "the usage")

File: docs/java-rest/high-level/migration.asciidoc [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+
+=== Motivations around a new Java client
+
+The existing `TransportClient` has been part of Elasticsearch since https://github.com/elastic/elasticsearch/blob/b3337c312765e51cec7bde5883bbc0a08f56fb65/modules/elasticsearch/src/main/java/org/elasticsearch/client/transport/TransportClient.java[its very first commit].
+Although it covers the whole set of Elasticsearch APIs, it always has had various flaws:
+
+- it is tightly coupled with Elasticsearch and uses Elasticsearch's internal implementations
+- it uses a home grown binary serialization to transport Java objects over the wire
+
+These two aspects forces any Java application that communicates with an Elasticsearch cluster to use
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: s/forces/force

File: docs/java-rest/high-level/migration.asciidoc [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+== Migration Guide
+
+This section describes how to migrate existing code from the `TransportClient`
+to the new Java High Level REST Client released with the version 5.6.0
+of Elasticsearch.
+
+=== Motivations around a new Java client
+
+The existing `TransportClient` has been part of Elasticsearch since https://github.com/elastic/elasticsearch/blob/b3337c312765e51cec7bde5883bbc0a08f56fb65/modules/elasticsearch/src/main/java/org/elasticsearch/client/transport/TransportClient.java[its very first commit].
+Although it covers the whole set of Elasticsearch APIs, it always has had various flaws:
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: This may make people think the the high level client doesn't depend on Elasticsearch, also we don't mention security and wanting to have REST a single entrypoing for a cluster. Shall we link https://www.elastic.co/blog/state-of-the-official-elasticsearch-java-clients that mentions all that and keep this explanation a bit shorter?

We could say:

The `TransportClient` has been part of Elasticsearch since its very first commit. It is a special client as it uses the transport protocol to communicate with Elasticsearch, which causes compatibility problems if the client is not on the same version as the Elasticsearch instances it talks to.

We released a low-level REST client in 2016, which is based on the well known Apache HTTP client and it allows to communicate with an Elasticsearch cluster in any version using HTTP. On top of that we released the high-level REST client which is based on the low-level client but takes care of request marshalling and response un-marshalling.

Have a look at (here the link) to know more about these changes.

File: docs/java-rest/high-level/migration.asciidoc [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+We decided to tackle these issues and we released a low-level REST client in 2016. This low-level REST client
+is based on the well-known Apache HTTP Client and it allows to communicate with an Elasticsearch cluster in any
+version using HTTP. On top of it, we released the high-level REST client. It is based on the low-level client
+and it provides a <<java-rest-high-supported-apis,subset of Elasticsearch's APIs>> that uses Elasticsearch core
+Java classes, like the `TransportClient` does, in order to perform requests and to parse responses.
+
+=== Prerequisite
+
+The Java High Level Rest Client requires Java `1.8`. It can be used to send requests
+to an Elasticsearch cluster that is in version 5.6.0 or above.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: we should probably more clear on whether it's fine to talk from 5.6.0 to 6.0.0. I guess it is, but new apis won't be known to the client this way. Also breaking changes to the API may cause unexpected results.

File: docs/java-rest/high-level/migration.asciidoc [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+
+Java application that uses the `TransportClient` declares the following dependency:
+
+["source","xml",subs="attributes"]
+--------------------------------------------------
+<dependency>
+    <groupId>org.elasticsearch.client</groupId>
+    <artifactId>transport</artifactId>
+    <version>...</version>
+</dependency>
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: I wonder if we have to put the whole maven dependency here, especially given that it's the old one. Maybe we can just say replace the o.e.client:transport dependency with... ?

File: docs/java-rest/high-level/migration.asciidoc [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+
+This dependency must be replaced by a new dependency on the
+high-level client. The <<java-rest-high-usage,Getting Started>> page shows
+ typical configurations for Maven and Gradle and presents the
+ <<java-rest-high-usage-dependencies, dependencies>> brought by the
+ high-level client.
+
+=== Changing the client's initialization code
+
+The `TransportClient` is typically initialized with something like:
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: s/with something like/as follows ?

File: docs/java-rest/high-level/migration.asciidoc [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+--------------------------------------------------
+
+The initialization of a `RestHighLevelClient` is different. It first requires the initialization
+of a <<java-rest-low-usage-initialization,low-level client>>:
+
+[source,java]
+--------------------------------------------------
+RestClient lowLevelRestClient = RestClient.builder(
+                new HttpHost("host", 9200, "http")).build();
+--------------------------------------------------
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: shall we highlight that we are using the 9200 port here instead of 9300 ?

File: docs/java-rest/high-level/migration.asciidoc [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+
+[source,java]
+--------------------------------------------------
+RestHighLevelClient client =
+    new RestHighLevelClient(lowLevelRestClient);
+--------------------------------------------------
+
+Both `RestClient` and `RestHighLevelClient` are thread safe. They are
+ typically instantiated by the application at startup time or when the
+ first request is executed, and must both be closed when they are not
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: the high level client doesn't need to be closed. I think we can add here a snippet that closes the transport client and say that it needs to be replaced with lowLevelClient.close ?

File: docs/java-rest/high-level/migration.asciidoc [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+More important, the high-level client:
+
+- does not provide indices or cluster management APIs. Management
+operations can be executed by external scripts or
+<<java-rest-high-level-migration-manage-indices, using the low-level client>>.
+- does not support request builders. The legacy methods like
+`client.prepareIndex()` must be changed to use
+ request constructors like `new IndexRequest()` to create requests
+ objects. The requests are then executed using synchronous or
+ asynchronous dedicated methods like `client.index()` or `client.indexAsync()`.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: can we invert the two bullet points, leaving request builders first?

File: docs/java-rest/high-level/migration.asciidoc [hunk 29]
```diff
@@ -0,0 +1,353 @@
...
+
+[[java-rest-high-level-migration-manage-indices]]
+==== Manage Indices using the Low-Level REST Client
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: I would move this paragraph to the end of the page

File: docs/java-rest/high-level/migration.asciidoc [hunk 30]
```diff
@@ -0,0 +1,353 @@
...
+
+The low-level client is able to execute any kind of HTTP requests, and can
+therefore be used to manage indices and retrieve information about the cluster.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: and can therefore be used to call the APIs that are not yet supported by the high level client.

File: docs/java-rest/high-level/migration.asciidoc [hunk 31]
```diff
@@ -0,0 +1,353 @@
...
+
+Once the `RestHighLevelClient` is initialized, it can then be used to
+execute any of the <<java-rest-high-supported-apis,supported APIs>>.
+
+=== Changing the application's code
+
+The `RestHighLevelClient` does not provide the same methods than
+the `TransportClient` to execute the requests.
+
+More important, the high-level client:
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: More importantly

File: docs/java-rest/high-level/migration.asciidoc [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+ typically instantiated by the application at startup time or when the
+ first request is executed, and must both be closed when they are not
+ needed anymore or when the application is stopped.
+
+Once the `RestHighLevelClient` is initialized, it can then be used to
+execute any of the <<java-rest-high-supported-apis,supported APIs>>.
+
+=== Changing the application's code
+
+The `RestHighLevelClient` does not provide the same methods than
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: The `RestHighLevelClient` supports the same request and response object as the `TransportClient`, but exposes slightly different methods to send the requests.

**Review by javanna — APPROVED** [review]
I added a few more comments, but indeed it is getting close. Great work @tlrx thanks a lot

File: docs/java-rest/high-level/migration.asciidoc (hunk 22 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: > This may make people think the the high level client doesn't depend on Elasticsearch,

Right, this can be a bit confusing and it's hard to resume all the motivations in a single paragraph.

I just borrow your test and let's go!

File: docs/java-rest/high-level/migration.asciidoc (hunk 26 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Good idea

File: docs/java-rest/high-level/migration.asciidoc (hunk 27 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: good point

File: docs/java-rest/high-level/migration.asciidoc (hunk 28 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Sure

File: docs/java-rest/high-level/migration.asciidoc (hunk 29 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Makes sense. Should I also move "Checking Cluster Health using the Low-Level REST Client" down?

File: docs/java-rest/high-level/migration.asciidoc (hunk 30 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Good point too

File: docs/java-rest/high-level/migration.asciidoc (hunk 29 above)
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: yes please

**Comment by tlrx** [comment]
Thanks @javanna - changes have been merged. I'll merge this unless you have more suggestions.

File: docs/java-rest/high-level/migration.asciidoc [hunk 33]
```diff
@@ -0,0 +1,365 @@
...
+client and it allows to communicate with an Elasticsearch cluster in any version using HTTP.
+On top of that we released the high-level REST client which is based on the low-level client
+but takes care of request marshalling and response un-marshalling.
+
+If you're interested in knowing more about these changes, we wrote a blog post about the
+https://www.elastic.co/blog/state-of-the-official-elasticsearch-java-clients[state of the official Elasticsearch Java clients].
+
+=== Prerequisite
+
+The Java High Level Rest Client requires Java `1.8`. It can be used to send requests
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: I would rephrase this:

The Java High Level Rest Client requires Java `1.8` and depends on Elasticsearch core. The client version is the same as the Elasticsearch version that the client was developed for. The High Level Client is backwards compatible but can only communicate with Elasticsearch version 5.5 and onwards. The High Level Client is forward compatible as well, meaning that it supports communicating with a later version of Elasticsearch than the one it was developed for. It is recommended to upgrade the High Level Client when upgrading the Elasticsearch cluster to a new major version, as REST API breaking changes may cause unexpected results, and newly added APIs will only be supported by the newer version of the client. The client should be updated last, once all of the nodes in the cluster have been upgraded.



I am now thinking if this should go somewhere else rather than in the migration guide.

File: docs/java-rest/high-level/migration.asciidoc [hunk 34]
```diff
@@ -0,0 +1,365 @@
...
+Adapting existing code to use the `RestHighLevelClient` instead of the `TransportClient`
+requires the following steps:
+
+- Update dependencies
+- Update client initialization
+- Update application code
+
+=== Updating the dependencies
+
+Java application that uses the `TransportClient` depend on the
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: depend**s**

File: docs/java-rest/high-level/migration.asciidoc [hunk 35]
```diff
@@ -0,0 +1,365 @@
...
+must be replaced by a new dependency on the high-level client.
+
+The <<java-rest-high-usage,Getting Started>> page shows
+ typical configurations for Maven and Gradle and presents the
+ <<java-rest-high-usage-dependencies, dependencies>> brought by the
+ high-level client.
+
+=== Changing the client's initialization code
+
+The `TransportClient` is typically initialized with something as follows:
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: typically initialized as follows

File: docs/java-rest/high-level/migration.asciidoc [hunk 36]
```diff
@@ -0,0 +1,365 @@
...
+
+[source,java]
+--------------------------------------------------
+RestClient lowLevelRestClient = RestClient.builder(
+                new HttpHost("host", 9200, "http")).build();
+--------------------------------------------------
+
+NOTE: The `RestClient` uses Elasticsearch's HTTP service which is
+ bounded by default on `9200`. This port is different from the port
+ used to connect to Elasticsearch with a `TransportClient`.
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: great!!!!

File: docs/java-rest/high-level/migration.asciidoc [hunk 37]
```diff
@@ -0,0 +1,365 @@
...
+Must be replaced with:
+
+[source,java]
+--------------------------------------------------
+lowLevelRestClient.close();
+--------------------------------------------------
+
+=== Changing the application's code
+
+The `RestHighLevelClient` supports the same request and response object
```
Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: object**s**

**Review by javanna — APPROVED** [review]
I left a couple of comments, but this looks great, thanks a lot

File: docs/java-rest/high-level/migration.asciidoc (hunk 35 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: oops sorry, it shouldn't have been reverted

File: docs/java-rest/high-level/migration.asciidoc (hunk 33 above)
Comment (tlrx) [inline docs/java-rest/high-level/migration.asciidoc]: Awesome. I added your text in a "Compatibility" section in the usage page. I created a link in the migration guide to this page.

Comment (javanna) [inline docs/java-rest/high-level/migration.asciidoc]: Sounds great thank you
